In [12]:
# Importing libraries
import pandas as pd
import numpy as np

In [25]:
# Loading adult data set
df = pd.read_csv('adult Data Set.csv', skipinitialspace = True)
df.head(50)

age         workclass  fnlwgt     education  education-num  \
0    39         State-gov   77516     Bachelors             13   
1    50  Self-emp-not-inc   83311     Bachelors             13   
2    38           Private  215646       HS-grad              9   
3    53           Private  234721          11th              7   
4    28           Private  338409     Bachelors             13   
5    37           Private  284582       Masters             14   
6    49           Private  160187           9th              5   
7    52  Self-emp-not-inc  209642       HS-grad              9   
8    31           Private   45781       Masters             14   
9    42           Private  159449     Bachelors             13   
10   37           Private  280464  Some-college             10   
11   30         State-gov  141297     Bachelors             13   
12   23           Private  122272     Bachelors             13   
13   32           Private  205019    Assoc-acdm             12   
14   40           Private  121772     Assoc-voc             11   
15   34           Private  245487       7th-8th              4   
16   25  Self-emp-not-inc  176756       HS-grad              9   
17   32           Private  186824       HS-grad              9   
18   38           Private   28887          11th              7   
19   43  Self-emp-not-inc  292175       Masters             14   
20   40           Private  193524     Doctorate             16   
21   54           Private  302146       HS-grad              9   
22   35       Federal-gov   76845           9th              5   
23   43           Private  117037          11th              7   
24   59           Private  109015       HS-grad              9   
25   56         Local-gov  216851     Bachelors             13   
26   19           Private  168294       HS-grad              9   
27   54                 ?  180211  Some-college             10   
28   39           Private  367260       HS-grad              9   
29   49           Private  193366       HS-grad              9   
30   23         Local-gov  190709    Assoc-acdm             12   
31   20           Private  266015  Some-college             10   
32   45           Private  386940     Bachelors             13   
33   30       Federal-gov   59951  Some-college             10   
34   22         State-gov  311512  Some-college             10   
35   48           Private  242406          11th              7   
36   21           Private  197200  Some-college             10   
37   19           Private  544091       HS-grad              9   
38   31           Private   84154  Some-college             10   
39   48  Self-emp-not-inc  265477    Assoc-acdm             12   
40   31           Private  507875           9th              5   
41   53  Self-emp-not-inc   88506     Bachelors             13   
42   24           Private  172987     Bachelors             13   
43   49           Private   94638       HS-grad              9   
44   25           Private  289980       HS-grad              9   
45   57       Federal-gov  337895     Bachelors             13   
46   53           Private  144361       HS-grad              9   
47   44           Private  128354       Masters             14   
48   41         State-gov  101603     Assoc-voc             11   
49   29           Private  271466     Assoc-voc             11   

           marital-status         occupation   relationship  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
5      Married-civ-spouse    Exec-managerial           Wife   
6   Married-spouse-absent      Other-service  Not-in-family   
7      Married-civ-spouse    Exec-managerial        Husband   
8           Never-married     Prof-specialty  Not-in-family   
9      

#### 1. Shape of dataset

In [26]:
df.shape

(32561, 15)

There are total 15 columns in the dataset with 32561 entries.

#### 2. Print column names 

In [27]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'Salary_range'],
      dtype='object')

#### 3. How many men and women (sex feature) are represented in this dataset? 

In [28]:
df.sex.value_counts()

Male      21790
Female    10771
Name: sex, dtype: int64

There are 21790 males and 10771 females in the dataset.

#### 4. What is the average age (age feature) of women? 

In [29]:
df.age.mean()

38.58164675532078

The average age of the adults in the dataset is 38.58 ~ 39

#### 5. What is the percentage of German citizens (native-country feature)?

In [32]:
((df['native-country'] == 'Germany').sum())/len(df)*100

0.42074874850281013

#### 6. What are the mean and standard deviation of age for those who earn more than 50K per year (salary feature) and those who earn less than 50K per year? 

##### Who earn more than 50K per year 

In [33]:
df[df['Salary_range'] == '>50K']['age'].mean()

44.24984058155847

In [34]:
df[df['Salary_range'] == '>50K']['age'].std()

10.51902771985177

##### Who earn less than 50K per year 

In [35]:
df[df['Salary_range'] == '<=50K']['age'].mean()

36.78373786407767

In [36]:
df[df['Salary_range'] == '<=50K']['age'].std()

14.020088490824813

#### 7. Is it true that people who earn more than 50K have at least high school education? (education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters or Doctorate feature) 

In [39]:
df[df['Salary_range'] == '>50K']['education'].unique()

array(['HS-grad', 'Masters', 'Bachelors', 'Some-college', 'Assoc-voc',
       'Doctorate', 'Prof-school', 'Assoc-acdm', '7th-8th', '12th',
       '10th', '11th', '9th', '5th-6th', '1st-4th'], dtype=object)

No, this is not true, as there are also people who earn more that 50K and lies in the category of educational qualification below high school.

#### 8. Display age statistics for each race (race feature) and each gender (sex feature). Use groupby() and describe(). Find the maximum age of men of Amer-Indian-Eskimo race. 

In [41]:
df.groupby(['race', 'sex'])[['age']].describe()

age                                          \
                             count       mean        std   min   25%   50%   
race               sex                                                       
Amer-Indian-Eskimo Female    119.0  37.117647  13.114991  17.0  27.0  36.0   
                   Male      192.0  37.208333  12.049563  17.0  28.0  35.0   
Asian-Pac-Islander Female    346.0  35.089595  12.300845  17.0  25.0  33.0   
                   Male      693.0  39.073593  12.883944  18.0  29.0  37.0   
Black              Female   1555.0  37.854019  12.637197  17.0  28.0  37.0   
                   Male     1569.0  37.682600  12.882612  17.0  27.0  36.0   
Other              Female    109.0  31.678899  11.631599  17.0  23.0  29.0   
                   Male      162.0  34.654321  11.355531  17.0  26.0  32.0   
White              Female   8642.0  36.811618  14.329093  17.0  25.0  35.0   
                   Male    19174.0  39.652498  13.436029  17.0  29.0  38.0   

                                        
                             75%   max  
race               sex                  
Amer-Indian-Eskimo Female  46.00  80.0  
                   Male    45.00  82.0  
Asian-Pac-Islander Female  43.75  75.0  
                   Male    46.00  90.0  
Black              Female  46.00  90.0  
                   Male    46.00  90.0  
Other              Female  39.00  74.0  
                   Male    42.00  77.0  
White              Female  46.00  90.0  
                   Male    49.00  90.0

Maximum age of men of Amer-Indian-Eskimo race is 82.

#### 9. Among whom is the proportion of those who earn a lot (>50K) greater: married or single men (marital-status feature)? Consider as married those who have a marital-status starting with Married (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), the rest are considered bachelors.

In [46]:
df[(df['sex'] == 'Male') & (df['marital-status'].str.startswith('Married'))]['Salary_range'].value_counts()

<=50K    7576
>50K     5965
Name: Salary_range, dtype: int64

In [48]:
df[(df['sex'] == 'Male') & ~(df['marital-status'].str.startswith('Married'))]['Salary_range'].value_counts()

<=50K    7552
>50K      697
Name: Salary_range, dtype: int64

The proportion of men who earn >50K are are married is greater than single men.

#### 10. What is the maximum number of hours a person works per week (hours-per-week feature)? How many people work such a number of hours, and what is the percentage of those who earn a lot (>50K) among them?

In [51]:
max_hours = df['hours-per-week'].max()
max_hours

99

The maximum number of hours a person works per week is 99 hours.

In [52]:
df[df['hours-per-week'] == max_hours]

age         workclass  fnlwgt     education  education-num  \
935     37           Private  176900       HS-grad              9   
1172    25           Private  404616       Masters             14   
1887    55  Self-emp-not-inc  184425  Some-college             10   
3578    37      Self-emp-inc  382802     Doctorate             16   
4086    50                 ?  174964          10th              6   
...    ...               ...     ...           ...            ...   
30992   40      Self-emp-inc  104235       Masters             14   
31681   43           Private  266324       HS-grad              9   
31699   43  Self-emp-not-inc  163108  Some-college             10   
32469   58      Self-emp-inc  181974     Doctorate             16   
32531   30                 ?   33811     Bachelors             13   

           marital-status       occupation   relationship                race  \
935    Married-civ-spouse     Craft-repair        Husband               White   
1172   Married-civ-spouse  Farming-fishing  Not-in-family               White   
1887   Married-civ-spouse  Farming-fishing        Husband               White   
3578   Married-civ-spouse   Prof-specialty        Husband               Black   
4086   Married-civ-spouse                ?        Husband               White   
...                   ...              ...            ...                 ...   
30992       Never-married    Other-service      Own-child               White   
31681  Married-civ-spouse     Craft-repair        Husband               White   
31699  Married-civ-spouse  Farming-fishing        Husband               White   
32469       Never-married   Prof-specialty  Not-in-family               White   
32531       Never-married                ?  Not-in-family  Asian-Pac-Islander   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
935      Male             0             0              99  United-States   
1172     Male             0             0              99  United-States   
1887     Male             0             0              99  United-States   
3578     Male             0             0              99  United-States   
4086     Male             0             0              99  United-States   
...       ...           ...           ...             ...            ...   
30992    Male             0             0              99  United-States   
31681    Male             0          1902              99  United-States   
31699    Male             0             0              99  United-States   
32469  Female             0             0              99              ?   
32531  Female             0             0              99  United-States   

      Salary_range  
935           >50K  
1172          >50K  
1887          >50K  
3578          >50K  
4086         <=50K  
...            ...  
30992        <=50K  
31681         >50K  
31699        <=50K  
32469        <=50K  
32531        <=50K  

[85 rows x 15 columns]

In [55]:
df[(df['hours-per-week'] == max_hours) & (df['Salary_range'] == '>50K')]

age         workclass  fnlwgt     education  education-num  \
935     37           Private  176900       HS-grad              9   
1172    25           Private  404616       Masters             14   
1887    55  Self-emp-not-inc  184425  Some-college             10   
3578    37      Self-emp-inc  382802     Doctorate             16   
4441    43      Self-emp-inc  286750   Prof-school             15   
8393    50      Self-emp-inc  119099  Some-college             10   
8796    39           Private   70995     Bachelors             13   
10986   38           Private  111499       HS-grad              9   
13558   38           Private  191103   Prof-school             15   
13858   37           Private   51264     Doctorate             16   
15460   37           Private  241174     Bachelors             13   
16604   73  Self-emp-not-inc  102510       7th-8th              4   
21147   39           Private  348521  Some-college             10   
22720   39           Private  237713   Prof-school             15   
23649   45           Private  390368  Some-college             10   
24483   39      Self-emp-inc  124685       Masters             14   
25713   32           Private  183304     Assoc-voc             11   
25831   31           Private  147284     Doctorate             16   
26596   35           Private  297574    Assoc-acdm             12   
29187   31           Private  178370       HS-grad              9   
29989   43         Local-gov  118853  Some-college             10   
30036   40           Private   25005     Bachelors             13   
30610   51  Self-emp-not-inc   24790  Some-college             10   
30766   42  Self-emp-not-inc   39539       HS-grad              9   
31681   43           Private  266324       HS-grad              9   

           marital-status        occupation   relationship  \
935    Married-civ-spouse      Craft-repair        Husband   
1172   Married-civ-spouse   Farming-fishing  Not-in-family   
1887   Married-civ-spouse   Farming-fishing        Husband   
3578   Married-civ-spouse    Prof-specialty        Husband   
4441   Married-civ-spouse    Prof-specialty        Husband   
8393   Married-civ-spouse   Exec-managerial        Husband   
8796   Married-civ-spouse  Transport-moving        Husband   
10986  Married-civ-spouse             Sales        Husband   
13558  Married-civ-spouse    Prof-specialty        Husband   
13858  Married-civ-spouse    Prof-specialty        Husband   
15460  Married-civ-spouse    Prof-specialty        Husband   
16604  Married-civ-spouse   Farming-fishing        Husband   
21147  Married-civ-spouse   Farming-fishing        Husband   
22720  Married-civ-spouse             Sales        Husband   
23649  Married-civ-spouse             Sales        Husband   
24483            Divorced   Exec-managerial  Not-in-family   
25713  Married-civ-spouse  Transport-moving        Husband   
25831  Married-civ-spouse    Prof-specialty        Husband   
26596  Married-civ-spouse   Exec-managerial        Husband   
29187            Divorced     Other-service      Unmarried   
29989  Married-civ-spouse      Tech-support        Husband   
30036  Married-civ-spouse    Prof-specialty        Husband   
30610  Married-civ-spouse   Exec-managerial        Husband   
30766  Married-civ-spouse   Farming-fishing        Husband   
31681  Married-civ-spouse      Craft-repair        Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
935                 White    Male             0             0              99   
1172                White    Male             0             0              99   
1887                White    Male             0             0              99   
3578                Black    Male             0             0              99   
4441                Black    Male             0             0              99   
8393                White    Male             0             0              99   
8796                White    Mal

#### 11. Count the average time of work (hours-per-week) for those who earn a little and a lot (salary) for each country (native-country). What will these be for Japan? 

In [61]:
countries = df['native-country'].unique()
# For those who earn less
for cou in countries:
    mean_hour = df[df['Salary_range'] == '<=50K']['hours-per-week'].mean()
    print(cou , " - " , mean_hour)

United-States  -  38.840210355987054
Cuba  -  38.840210355987054
Jamaica  -  38.840210355987054
India  -  38.840210355987054
?  -  38.840210355987054
Mexico  -  38.840210355987054
South  -  38.840210355987054
Puerto-Rico  -  38.840210355987054
Honduras  -  38.840210355987054
England  -  38.840210355987054
Canada  -  38.840210355987054
Germany  -  38.840210355987054
Iran  -  38.840210355987054
Philippines  -  38.840210355987054
Italy  -  38.840210355987054
Poland  -  38.840210355987054
Columbia  -  38.840210355987054
Cambodia  -  38.840210355987054
Thailand  -  38.840210355987054
Ecuador  -  38.840210355987054
Laos  -  38.840210355987054
Taiwan  -  38.840210355987054
Haiti  -  38.840210355987054
Portugal  -  38.840210355987054
Dominican-Republic  -  38.840210355987054
El-Salvador  -  38.840210355987054
France  -  38.840210355987054
Guatemala  -  38.840210355987054
China  -  38.840210355987054
Japan  -  38.840210355987054
Yugoslavia  -  38.840210355987054
Peru  -  38.840210355987054
Outl

In [63]:
# For those who earn a lot
for cou in countries:
    mean_hour = df[df['Salary_range'] == '>50K']['hours-per-week'].mean()
    print(cou , " - " , mean_hour)

United-States  -  45.473026399693914
Cuba  -  45.473026399693914
Jamaica  -  45.473026399693914
India  -  45.473026399693914
?  -  45.473026399693914
Mexico  -  45.473026399693914
South  -  45.473026399693914
Puerto-Rico  -  45.473026399693914
Honduras  -  45.473026399693914
England  -  45.473026399693914
Canada  -  45.473026399693914
Germany  -  45.473026399693914
Iran  -  45.473026399693914
Philippines  -  45.473026399693914
Italy  -  45.473026399693914
Poland  -  45.473026399693914
Columbia  -  45.473026399693914
Cambodia  -  45.473026399693914
Thailand  -  45.473026399693914
Ecuador  -  45.473026399693914
Laos  -  45.473026399693914
Taiwan  -  45.473026399693914
Haiti  -  45.473026399693914
Portugal  -  45.473026399693914
Dominican-Republic  -  45.473026399693914
El-Salvador  -  45.473026399693914
France  -  45.473026399693914
Guatemala  -  45.473026399693914
China  -  45.473026399693914
Japan  -  45.473026399693914
Yugoslavia  -  45.473026399693914
Peru  -  45.473026399693914
Outl

#### For Japan

In [64]:
# For those who earn less
df[(df['Salary_range'] == '<=50K') & (df['native-country'] == 'Japan')]['hours-per-week'].mean()

41.0

In [65]:
# For those who earn a lot
df[(df['Salary_range'] == '>50K') & (df['native-country'] == 'Japan')]['hours-per-week'].mean()

47.958333333333336